File: nexml/javascript/Phylo/Taxa/Taxon.js
Phylo.Taxa.Taxon = function (args) {
	if (args==null) args = {};	
	args["tag"] = "otu";
    this.XMLWritable(args);
    this._type      = Phylo.Util.CONSTANT._TAXON_;
    this._container = Phylo.Util.CONSTANT._TAXA_;
    return this;
}
copyPrototype(Phylo.Taxa.Taxon,Phylo.Util.XMLWritable);

Phylo.Taxa.Taxon.prototype.set_data = function (datum) {
	if ( looks_like_object( datum, Phylo.Util.CONSTANT._DATUM_ ) ) {
		Phylo.Mediators.TaxaMediator.set_link({
			'one'  : this,
			'many' : datum
		});
	}
	return this;
};

Phylo.Taxa.Taxon.prototype.set_nodes = function (node) {
	if ( looks_like_object( node, Phylo.Util.CONSTANT._NODE_ ) ) {
		Phylo.Mediators.TaxaMediator.set_link({
			'one'  : this,
			'many' : node
		});
	}
	return this;
};

Phylo.Taxa.Taxon.prototype.unset_datum = function (datum) {
	Phylo.Mediators.TaxaMediator.remove_link({
		'one'  : this,
		'many' : datum
	});
	return this;
};

Phylo.Taxa.Taxon.prototype.unset_node = function (node) {
	Phylo.Mediators.TaxaMediator.remove_link({
		'one'  : this,
		'many' : node
	});
	return this;
};

Phylo.Taxa.Taxon.prototype.get_data = function () {
    return Phylo.Mediators.TaxaMediator.get_link({
        'source' : this, 
        'type'   : Phylo.Util.CONSTANT._DATUM_()
    });
};

Phylo.Taxa.Taxon.prototype.get_nodes = function () {
    return Phylo.Mediators.TaxaMediator.get_link({
        'source' : this, 
        'type'   : Phylo.Util.CONSTANT._NODE_()
    });
};

Phylo.Taxa.Taxon.prototype._type = function() {
    return this._type;
};

Phylo.Taxa.Taxon.prototype._container = function() {
    return this._container;
};

File: java/src/org/biophylo/Taxa/Taxon.java
package org.biophylo.Taxa;
import org.biophylo.Util.*;
import org.biophylo.*;
import org.biophylo.Forest.*;
import org.biophylo.Matrices.*;
import org.biophylo.Mediators.ObjectMediator;
import org.biophylo.Mediators.TaxaMediator;
import java.util.*;
public class Taxon extends Containable {
	private static TaxaMediator taxaMediator = TaxaMediator.getInstance();
	public Taxon () {
		this.type = CONSTANT.TAXON;
		this.container = CONSTANT.TAXA;
		this.tag = "otu";
	}
	public Node[] getNodes () {
		int taxonId = this.getId();
		Vector tl = taxaMediator.getLink(taxonId, CONSTANT.NODE);
		Node[] result = new Node[tl.size()];
		tl.copyInto(result);
		return result;
	}
	public void setNodes(Node node) {
		int taxonId = this.getId();
		int linkerId = node.getId();
		taxaMediator.setLink(taxonId, linkerId);
	}
	public void unsetNode(Node node) {
		int taxonId = this.getId();
		int linkerId = node.getId();
		taxaMediator.removeLink(taxonId, linkerId);		
	}	
	public Datum[] getData() {
		int taxonId = this.getId();
		Vector tl = taxaMediator.getLink(taxonId, CONSTANT.DATUM);
		Datum[] result = new Datum[tl.size()];
		tl.copyInto(result);
		return result;		
	}
	public void setData(Datum datum) {
		int taxonId = this.getId();
		int linkerId = datum.getId();
		taxaMediator.setLink(taxonId, linkerId);		
	}
	public void unsetDatum(Datum datum) {
		int taxonId = this.getId();
		int linkerId = datum.getId();
		taxaMediator.removeLink(taxonId, linkerId);			
	}
	protected void finalize() throws Throwable {
		taxaMediator.removeLink(this.getId(), -1);
	  //do finalization here
	  super.finalize(); //not necessary if extending Object.
	} 

}
